custom_components/control4/__init__.py

```diff
@@ -59,6 +60,18 @@
     Platform.LOCK,
 ]
 
+async def call_c4_api_retry(func, *func_args):
+     """Call C4 API function and retry on failure."""
+     for i in range(API_RETRY_TIMES):
+         try:
+             output = await func(*func_args)
+             return output
+         except BadCredentials as exception:
+             raise ConfigEntryAuthFailed(exception) from exception
+         except client_exceptions.ClientError as exception:
+             _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+             if i == API_RETRY_TIMES - 1:
```

Comment:
Is there a reason why we need to add our own limit to the number of retries? I guess this is because HA doesn't have its own way of limiting the retries? 

Even so, it seems like it would be better to have a progressively increasing backoff instead of giving up forever after 5 tries.

Reply:
Yeah, I’m not super familiar with the internals of HA so I’m not certain they don’t have something built in for this but just assumed it didn’t exist since the core PR didn’t use it. 

Generally speaking though I agree about having a backoff instead of max attempts. I had originally [made a change](https://github.com/mellis/hass-control4/commit/905b9aa92e9344ed474c377fabd9cc968a86c92c) like that in my fork but I wasn’t sure how much we wanted the hacs version to deviate from the core one so I tried to more directly back port the core change here. What’s the long term plan with this repo and would you rather use the backoff style retry like I originally wrote?

I’m on vacation at the moment but will probably have time to update this PR sometime later next week.

Reply:
I looked into this a bit more and HA does automatically retry if the failure occurs _during initial setup (HA startup)_, but wouldn't retry if the failure occurs after a successful setup - this is why the integration fails to refresh the token after startup, but doesn't fail to get the token during startup. https://developers.home-assistant.io/docs/integration_setup_failures/

Reply:
Instead of scheduling our own retries, we should use HA's built in scheduling functionality, like what we already use to schedule the token refreshes in the first place: https://github.com/lawtancool/hass-control4/blob/146464bb36a0283088226d0c0e730a0c4335da81/custom_components/control4/__init__.py#L240-L244